docs/source/troubleshooting/tracing.rst

```diff
@@ -0,0 +1,106 @@
+==================================================
+Tracing on NI Linux Real-Time
+==================================================
+
+.. contents:: Table of Contents
+   :depth: 2
+   :local:
+
+Introduction
+============
+
+Often times when debugging latency issues, race conditions or hangs on real-time embedded targets it is useful to know what the kernel/OS was doing just prior to the error condition. Acquiring and analyzing a kernel event or function trace can provide insights into latency spikes, race conditions, lock contention, priority inversions, missed loop iterations and so on.
+
+On NI Linux Real-Time (NILRT) there are several higher level tools that can be used for this like the "LabVIEW Execution Trace Toolkit". Documentation for using those tools is available elsewhere. This document only covers using the lower level APIs for interacting with the kernel ftrace system.
+
+ftrace
+======
+
+On NILRT systems the kernel debugfs is mounted at `/sys/kernel/debug`. The ftrace subsystem can be controlled via pseudo-files at `/sys/kernel/debug/tracing`. For a quick description of the configuration knobs available, see `/sys/kernel/debug/tracing/README`.
+
+Below is the typical sequence to start an *event trace* from a ssh/serial terminal prompt. The same operations can be performed via file read/writes from other programming languages including LabVIEW.
+
+.. note::
+   Files under `/sys/kernel/debug/tracing` are text files, use appropriate LabVIEW primitives when accessing them.
+
+.. note::
+   Tracing is a privileged operation. On older versions of NILRT, file ownership and/or permissions have to be adjusted when used from LabVIEW; e.g., with `chown -R lvuser:ni /sys/kernel/debug/tracing`. This change will not persist across a reboot.
+
+#. Stop any ongoing tracing and clear the buffer
+
+   .. code:: text
+
+      echo 0 > /sys/kernel/debug/tracing/tracing_on
+      echo 0 > /sys/kernel/debug/tracing/trace
+
+#. Enable tracing for all events
+
+   .. code:: text
+
+      echo 1 > /sys/kernel/debug/tracing/events/enable
+
+#. Start tracing
+
+   .. code:: text
+
+      echo 1 > /sys/kernel/debug/tracing/tracing_on
+
+#. (optional) Add custom markers to the trace
+
+   Custom markers can be added into the trace to track the application progress by writing into the trace_marker file:
+
+   .. code:: text
+
+      echo "something happened" > /sys/kernel/debug/tracing/trace_marker
+
+   Use this with caution and sparingly. Writing to the trace buffer adds additional latency.
+
+Analyzing the results
+=====================
+
+The resulting trace can be extracted by reading the `/sys/kernel/debug/tracing/trace` file directly. This will output a human readable textual representation. However this is usually pretty slow especially on large traces.
```

Comment:
Instead of "However this is usually pretty slow especially on large traces.", what about...

Reading the trace can be a slow operation, if it is large.